src/useSubtitles.js

```diff
@@ -0,0 +1,114 @@
+  const memoizedCurrentMetadata = React.useMemo(() => currentMetadata[language], [currentMetadata, language]);
+  const memoizedCurrentSubtitle = React.useMemo(() => currentSubtitle[language], [currentSubtitle, language])
+
+  const ref = useCallbackRef({
+    onMount: onMountEvent,
+    onUnmount: onUnmountEvent
+  })
+
+  /** @param {HTMLMediaElement} node */
+  function handleInitialLoad(node) {
+    const tracks = toIterable(node?.textTracks)
+
+    // Set initial subtitles
+    node?.addEventListener('loadedmetadata', () => {
+      tracks.forEach((x) => {
+        const hasSubtitles = subtitles[language].length
+        const hasMetadata = metadata[language].length
+
+        if (!hasSubtitles && x.kind === 'subtitles') {
+          setInitialSubtitles(x)
+        }
+        
+        if (!hasMetadata && x.kind === 'metadata') {
+          setInitialMetadata(x)
+        }
+      })
+    })
+
+    // Handle cue changes
+    tracks.forEach((x) => {
+      if (x.kind === 'subtitles') {
+        x.addEventListener("cuechange", onCueChangeEvent);
+      }
+      
+      if (x.kind === 'metadata') {
+        x.addEventListener("cuechange", onMetadataChangeEvent);
+      }
+
+      x.mode = "hidden";
+    })
+  }
+  
+  /** @param {HTMLMediaElement} node */
+  function handleUnmount(node) {
+    toIterable(node?.textTracks).forEach((x) => {
+      x.removeEventListener("cuechange", onCueChangeEvent);
+    })
+  }
+
+  /** @param {{ language: string, cues: TextTrackCueList }} _ */
+  function setInitialSubtitles({ cues, language }) {
+    if (!subtitles.length && cues) {
+      setSubtitles(x => ({ ...x, [language]: toIterable(cues) }));
+    }
+  }
+
+  /** @param {{ language: string, cues: TextTrackCueList }} _ */
+  function setInitialMetadata({ cues, language }) {
+    if (!metadata.length && cues) {
+      setMetadata(x => ({ ...x, [language]: toIterable(cues) }));
+    }
+  }
+
+  /** @param {{ target: { language: string, activeCues: TextTrackCueList } }} _ */
+  function handleMetadataChange({ target }) {
+    const [cue] = toIterable(target.activeCues).map(x => ({
+      text: isJSON(x.text) ? JSON.parse(x.text) : x.text,
+      startTime: x.startTime,
+      endTime: x.endTime
+    }))
+
+    setcurrentMetadata(x => ({ ...x, [target.language]: cue ?? initial }))
+  }
+ 
+  /** @param {{ target: { language: string, activeCues: TextTrackCueList } }} _ */
+  function handleCueChange({ target }) { 
+    const [cue] = toIterable(target.activeCues).map((x) => ({
+      text: x.getCueAsHTML().textContent,
+      voice: getVoiceFromCue(x.text),
+      startTime: x.startTime,
+      endTime: x.endTime
+    }))
+
+    setCurrentSubtitle(x => ({ ...x, [target.language]: cue ?? initialCurrentSubtitle }))
+  }
+
+  return {
+    subtitles: memoizedSubtitles,
+    metadata: memoizedMetadata,
+    current: {
+      subtitle: memoizedCurrentSubtitle,
+      metadata: memoizedCurrentMetadata
+    },
```

Comment:
Tricky, maar `current` is hier nog steeds iedere keer een nieuw object. Waarschijnlijk is het geen issue omdat er meestal destructuring gebruikt zal worden. 

Ik denk dat je alle memo's hier niet nodig hebt, omdat je met setState's werkt, dus tussen setState's door veranderen deze waardes sowieso al niet.